### PR TITLE
Implement basic Start, Stop, Status

### PR DESCRIPTION
**Pull request by romanlevin**
See `README.md` for functional examples.

## What's missing?

* No transport encryption
* No auth
* No log capture
* Client is a very crude CLI tool
* Automated tests

## What I learned?

Calling `.Wait` on `exec.Cmd` is necessary to populate the `Cmd` struct when the process exits.

Killed processes don't always have a populated `ProcessState`.

Probably want to store jobs by value instead of by reference, to make sure all updates are controlled by locks. Will look into this with next PR.

### CODE REVIEW

File: client/main.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	if err != nil {
+		log.Fatalf("did not connect: %v", err)
+	}
+	defer conn.Close()
+	c := jobs.NewJobsClient(conn)
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	switch command := os.Args[1]; command {
```
Comment (bernardjkim) [inline client/main.go:30]: Might want to add some bounds checks for args.

File: client/main.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+	case "run":
+		runJob(c, ctx)
```
Comment (xacrimon) [inline client/main.go]: context should be the first argument as standard in go

File: client/main.go [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+	case "status":
+		statusJob(c, ctx)
```
Comment (xacrimon) [inline client/main.go]: context should be the first argument as standard in go

File: client/main.go [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+	case "stop":
+		stopJob(c, ctx)
```
Comment (xacrimon) [inline client/main.go]: context should be the first argument as standard in go

File: client/main.go [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+	address = "localhost:12345"
+)
+
+func main() {
+	conn, err := grpc.Dial(address, grpc.WithInsecure(), grpc.WithBlock())
+	if err != nil {
+		log.Fatalf("did not connect: %v", err)
+	}
+	defer conn.Close()
+	c := jobs.NewJobsClient(conn)
```
Comment (bernardjkim) [inline client/main.go]: nit: Even though this is not production code, please use clear variable names for easier readability.

File: server/main.go [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+)
+
+func main() {
+	lis, err := net.Listen("tcp", ":12345")
+
+	if err != nil {
+		log.Fatalf("failed to listen: %v", err)
+	}
+
+	s := grpc.NewServer()
```
Comment (bernardjkim) [inline server/main.go]: How would you go about handling graceful shutdown of the server?

File: server/job.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
+package main
+
+import (
+	"github.com/google/uuid"
+	pb "github.com/romanlevin/felek/jobs"
+	"os/exec"
+)
+
+type job struct {
+	id        string
```
Comment (xacrimon) [inline server/job.go]: What's the format of id? What standard is used?

Comment (xacrimon) [reply server/job.go]: I see that it's an UUID later but please document weakly string variables if their format and intention isn't immediately obvious.

File: jobs/felek.proto [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+message JobStatus {
+    JobID id = 1; // A unique ID
+    oneof jobState {
+        RunningJob runningJob = 2;
+        StoppedJob stoppedJob = 3;
+    }
+}
+
+message RunningJob {
+    // XXX: Don't know if we only want this available for running processes,
```
Comment (xacrimon) [inline jobs/felek.proto:35]: The PID might be useful in some cases but it's still fairly niche. It's fine to leave the message empty as a status indicator too.

File: server/job.go [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+	cmd       *exec.Cmd
+	stopped   bool
+	exitError string
```
Comment (xacrimon) [inline server/job.go]: Please add a comment about the meaning and contents of this.

**Review by xacrimon — APPROVED** [review]

File: server/server.go [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+	if !ok {
+		return nil, fmt.Errorf("no such job id %v", jobID)
+	}
+
+	err := j.cmd.Process.Kill()
+	if err != nil {
+		return jobStatus(j), fmt.Errorf("failed to kill: %w", err)
+	}
+
+	s.jobsLock.Lock()
```
Comment (bernardjkim) [inline server/server.go]: Locking the entire server doesn't seem like the most optimal way to update a single job. Have you considered a separate lock per job?

File: server/job.go [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+		return &pb.JobStatus{
+			Id: &pb.JobID{Value: j.id},
+			JobState: &pb.JobStatus_StoppedJob{
+				StoppedJob: &pb.StoppedJob{
+					Exit: &pb.StoppedJob_ExitCode{
+						ExitCode: int64(c.ProcessState.ExitCode()),
+					},
+					SystemTime: int64(c.ProcessState.SystemTime()),
+					UserTime:   int64(c.ProcessState.UserTime()),
+					Stopped:    j.stopped,
```
Comment (bernardjkim) [inline server/job.go]: Is it safe to access the `stopped` field without acquiring a lock?

File: client/main.go [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
```
Comment (r0mant) [inline client/main.go]: Hmm, what's this timeout of 1 second used for?

File: client/main.go [hunk 13]
```diff
@@ -0,0 +1,81 @@
...
+	case "stop":
+		stopJob(c, ctx)
+	default:
+		log.Fatalf("must run with %s [run|status|stop] ..., instead got %s", os.Args[0], os.Args[1])
+	}
+}
+
+func runJob(c jobs.JobsClient, ctx context.Context) {
+	request := &jobs.JobStartRequest{
+		Path: os.Args[2],
```
Comment (r0mant) [inline client/main.go:48]: This can panic.

File: client/main.go [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+		Args: os.Args[3:],
+	}
+	log.Printf("sending request %v", request)
+	r, err := c.Start(ctx, request)
+	if err != nil {
+		log.Fatalf("could not start: %#v", err.Error())
```
Comment (r0mant) [inline client/main.go]: As a general pattern, I would have all functions return errors back to the caller, all the way up to the main, and log once/exit there.

File: client/main.go [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+	}
+	log.Printf("Job name: %v", r)
+}
+
+func stopJob(c jobs.JobsClient, ctx context.Context) {
+	id := &jobs.JobID{Value: os.Args[2]}
```
Comment (r0mant) [inline client/main.go:63]: This can panic.

File: client/main.go [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+	status, err := c.Stop(ctx, id)
+	if err != nil {
+		log.Fatalf("could not kill: %#v", err.Error())
+	}
+	logJobStatus(status)
+}
+
+func statusJob(c jobs.JobsClient, ctx context.Context) {
+	id := &jobs.JobID{Value: os.Args[2]}
```
Comment (r0mant) [inline client/main.go:75]: This can panic.

File: client/main.go [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+}
+
+func logJobStatus(status *jobs.JobStatus) {
+	switch jobState := status.JobState.(type) {
+	case *jobs.JobStatus_StoppedJob:
+		log.Printf("stopped job(%s): %v, stopped by user: %t", status.Id.Value, jobState.StoppedJob, jobState.StoppedJob.Stopped)
+	case *jobs.JobStatus_RunningJob:
+		log.Printf("running job(%s): %v", status.Id.Value, jobState.RunningJob)
+	default:
+		log.Fatalf("JobStatus.JobState has unexpected type %T", jobState)
```
Comment (r0mant) [inline client/main.go]: Return the error and let the caller decide how to react on it.

File: server/job.go [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+type job struct {
+	id        string
+	cmd       *exec.Cmd
+	stopped   bool
+	exitError string
+	owner     string
+}
+
+func jobStatus(j *job) *pb.JobStatus {
+	c := j.cmd
```
Comment (r0mant) [inline server/job.go]: Please use meaningful variable names.

File: server/job.go [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+	if j.exitError == "" && c.ProcessState != nil {
```
Comment (r0mant) [inline server/job.go]: Can this race?

File: server/job.go [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+					Exit: &pb.StoppedJob_ExitCode{
+						ExitCode: int64(c.ProcessState.ExitCode()),
+					},
+					SystemTime: int64(c.ProcessState.SystemTime()),
+					UserTime:   int64(c.ProcessState.UserTime()),
+					Stopped:    j.stopped,
+				},
+			},
+		}
+	} else if c.ProcessState == nil {
```
Comment (r0mant) [inline server/job.go]: Can this race?

File: server/server.go [hunk 21]
```diff
@@ -0,0 +1,115 @@
...
+	pb.UnimplementedJobsServer
+	jobsLock *sync.RWMutex
+	jobs     map[string]*job
+	logDir   string
+}
+
+func newServer() *server {
+	logDir, err := ioutil.TempDir("", "felek-logs-*")
+	if err != nil {
+		panic(err)
```
Comment (r0mant) [inline server/server.go]: Do not panic in server.

File: server/server.go [hunk 22]
```diff
@@ -0,0 +1,115 @@
...
+func (s *server) Stop(ctx context.Context, id *pb.JobID) (*pb.JobStatus, error) {
+	jobID := id.Value
+	j, ok := s.getJob(jobID)
+	if !ok {
+		return nil, fmt.Errorf("no such job id %v", jobID)
+	}
+
+	err := j.cmd.Process.Kill()
+	if err != nil {
+		return jobStatus(j), fmt.Errorf("failed to kill: %w", err)
```
Comment (r0mant) [inline server/server.go]: Why does this return `jobStatus(j), error` instead of `nil, err`?

File: server/job.go [hunk 23]
```diff
@@ -0,0 +1,68 @@
...
+		// Job still running
+		return &pb.JobStatus{
+			Id: &pb.JobID{Value: j.id},
+			JobState: &pb.JobStatus_RunningJob{
+				RunningJob: &pb.RunningJob{Pid: int64(j.cmd.Process.Pid)},
+			},
+		}
+	} else {
```
Comment (r0mant) [inline server/job.go]: It's pretty hard to follow the logic in this function due to verbosity and branching. I'd try to restructure it e.g. create `pb.JobStatus` once and then fill appropriate fields depending on the current job status with explanations of each case.

File: client/main.go (hunk 12 above)
Comment (romanlevin) [reply client/main.go]: Honestly I just assumed this timeout would cancel the gRPC call if it takes longer than a second, but testing it out with a dead server shows that's definitely not the case. Changed it to a background context for now, will look into it later.

File: server/job.go (hunk 23 above)
Comment (romanlevin) [reply server/job.go]: @r0mant please take a look at the current implementation. I'm not super happy with it still, but Go's way of handling `oneof` fields is inherently a bit awkward.

File: server/server.go (hunk 10 above)
Comment (romanlevin) [reply server/server.go]: That _is_ a much better idea given that we only ever write to the map once for each job.

File: server/main.go (hunk 6 above)
Comment (romanlevin) [reply server/main.go]: Implemented `SIGINT` handling and graceful shutdown in 74fc139.

File: server/job.go (hunk 11 above)
Comment (romanlevin) [reply server/job.go]: Definitely is not. Fixing as part of moving to per-job locks.

File: server/job.go (hunk 19 above)
Comment (romanlevin) [reply server/job.go]: I'm not entirely sure if it can race here, but I'm moving to a per-job lock as per @bernardjkim's suggestion, and there we will acquire a read lock in `jobStatus`, and a write lock when stopping a job, which should prevent races here.

File: server/server.go (hunk 22 above)
Comment (romanlevin) [reply server/server.go]: It should not.

File: server/job.go (hunk 20 above)
Comment (romanlevin) [reply server/job.go]: See https://github.com/romanlevin/felek/pull/2#discussion_r592320068

**Review by bernardjkim — APPROVED** [review]

File: server/job.go [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+	"sync"
+)
+
+type job struct {
+	id        string    // A UUID string
+	cmd       *exec.Cmd // The underlying Cmd object wrapping the job process
+	stopped   bool      // Has this job been stopped by a client?
+	exitError error     // The error returned by Wait on the Cmd
+	owner     string    // The username of the user who started the job
+	lock      *sync.RWMutex
```
Comment (r0mant) [inline server/job.go]: Mutex doesn't need to be a pointer here, as long as your functions have pointer receivers.

File: server/job.go [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+// exitedWithExitCode assumes a lock is held
+func (j *job) exitedWithExitCode() bool {
```
Comment (r0mant) [inline server/job.go]: Nit: I'd name these `exitedWithExitCodeLocked()` and `runningLocked()` to indicate they're supposed to be invoked under lock.

File: server/job.go [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+		},
+	}
+	return status
+}
+
+// newJob creates a job struct for a new Cmd, assigning it a UUID
+func newJob(cmd *exec.Cmd, owner string) *job {
+	id := uuid.NewString()
+	return &job{
+		id:    id,
```
Comment (r0mant) [inline server/job.go]: Nit: Just do `id: uuid.NewString()`, there's no need for extra variable really.

File: server/server.go [hunk 27]
```diff
@@ -0,0 +1,89 @@
...
+
+func (s *server) storeJob(j *job) {
+	s.jobs[j.id] = j
+}
+
+func (s *server) getJob(id string) (*job, bool) {
+	j, ok := s.jobs[id]
+	return j, ok
+
+}
```
Comment (r0mant) [inline server/server.go]: Can't these be called concurrently by multiple clients?

File: server/job.go [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+func (j *job) exitedWithExitCode() bool {
+	return j.exitError == nil && j.cmd.ProcessState != nil
+}
+
+// running assumes a lock is held
+func (j *job) running() bool {
+	return j.exitError == nil && j.cmd.ProcessState == nil
+}
+
+func (j *job) stop() error {
```
Comment (r0mant) [inline server/job.go:31]: What happens if multiple clients try to stop the job simultaneously?

File: server/server.go (hunk 27 above)
Comment (romanlevin) [reply server/server.go]: Yes. I'm protecting access to attributes, not to the job struct itself.

I guess an alternative would be to store job by value, and requiring a lock to put it back, but Cmd has to be stored by reference anyway.

I could also put the job struct in its own package and force the server to access it through methods only.

File: server/job.go (hunk 28 above)
Comment (romanlevin) [reply server/job.go:31]: One would get the lock and kill it, the others would error out trying to kill a killed process, or succeed in sending the signal if the process is still alive by that time.

I could 'Wait()' here to make sure the process is dead when the lock is released, but then it might complicate things with 'wait', not sure what happens with multiple waiters.

Comment (romanlevin) [reply server/job.go:31]: Turns out calling `.Wait()` on a finished processes returns an error (`"exec: Wait was already called"`), but in any case, I think `Stop` returning an error for all but one of multiple clients is working as intended.

File: server/server.go (hunk 27 above)
Comment (r0mant) [reply server/server.go]: I don't think it's a question about reference vs value, but rather whether the map access is concurrency-safe?

File: server/job.go (hunk 28 above)
Comment (r0mant) [reply server/job.go:31]: Any way to handle this more gracefully and prevent clients besides the first one try to kill the job?

Comment (romanlevin) [reply server/job.go:31]: Yeah, I brought it up earlier in this thread. We can `Wait` in `Stop`, before releasing the lock.

Another option is to check `stopped` before calling `Kill`.

File: server/server.go (hunk 27 above)
Comment (romanlevin) [reply server/server.go]: Oh I misunderstood the question earlier. 

Turns out the answer is "no, [it is not](https://golang.org/doc/faq#atomic_maps)". For a language so focused on concurrency I'm constantly surprised by decisions around it.

Guess I'm bringing back that server-wide lock to protect map access.

File: server/job.go (hunk 28 above)
Comment (romanlevin) [reply server/job.go:31]: I implemented the latter.

File: server/server.go (hunk 27 above)
Comment (romanlevin) [reply server/server.go]: Protected access to `jobs` with a `sync.RWMutex`.

**Review by r0mant — APPROVED** [review]
